### DISCUSSION TITLE
why spa mode still use hydration？

### DISCUSSION THREAD

**Discussion by SteveSuv** [remix-run/remix#9008] [Q&A] [answered]
spa mode doesn't has server, why still use `hydrateRoot` by default but not `createRoot`, I don't understand.

**Comment by alexanderson1993** [comment, marked as answer]
SPA mode still needs a way to generate the initial `index.html` that is first loaded when visitors come to your app. `hydrateRoot` is what is rendered into that `index.html`

**Reply by SteveSuv** [reply]
Thanks for your reply. But I tried toggle `hydrateRoot` to  `createRoot` in `entry.client.tsx`, it works fine too. So what's the difference between  two way?

**Reply by alexanderson1993** [reply]
Probably not a huge difference either way. With `createRoot`, React will throw out all of the initial HTML on the page and replace it with the React components. With `hydrateRoot`, React will keep the existing elements and associate them with their corresponding React elements. Since Remix still generates initial HTML based on your root route, I think `hydrateRoot` leads to faster app initialization and lower memory usage at first, but you're probably fine using `createRoot`.


What difference does it make to you? Why are you investigating the differences between `createRoot` and `hydrateRoot`?

**Reply by SteveSuv** [reply]
Because of `HydrateFallback` component.
when I want to use client vars like `window`,`location`,`document`, it shows `undefined`.
and I can't render client component in `HydrateFallback`, like `react-spinners`. remix will throw error.
what I don't understand is why I can't use client things in a pure client spa mode?
so I doubt the reason is `hydrateRoot `, but if I change to `createRoot`, it still throw error.🥹

```tsx
export function HydrateFallback() {
  console.log(window);
  return <div>{window.location.href} is loading...</div>;
}
```
```sh
ReferenceError: window is not defined
```
---
```tsx
import HashLoader from "react-spinners/HashLoader";

export function HydrateFallback() {
  return <HashLoader/>
}
```
```sh
Warning: React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: object.

Check your code at root.tsx:53.
at HydrateFallback
```

**Reply by alexanderson1993** [reply]
It's not strictly a pure client SPA. The initial `index.html` is compiled by Remix in a Node environment, where `window` doesn't exist. You need to wrap your window use in a little check:

```ts
export function HydrateFallback() {
  if (typeof window !== "undefined") {
    console.log(window)
  }
  // ...
}
```

I don't know why `<HashLoader>` wasn't working properly for you.

**Reply by SteveSuv** [reply]
Thanks,I understand
